### ISSUE TITLE
Offer enhanced multi-data iterator solutions

### ISSUE THREAD

**Issue by ninginthecloud** [PyTorchLightning/pytorch-lightning#9046]
### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:

https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
Review Lightning architecture & API #7740

As discussed in the above overview doc, there are some opportunities to improve support for multiple dataloaders. 
In current PyTorch Lightning, how to handle mult-dataloaders varies between train vs. val/test/predict:
- If you pass multiple dataloaders for training, these are combined into a single batch inside of the trainer.
- If you pass multiple dataloaders for evaluation/testing, Lightning sequentially loops through each loader. If anything, it's much easier for users to sequentially call Trainer.validate/test/predict(...) for each individual dataloader in a loop instead of doing this altogether. They can also do this in parallel, leading to speedups in overall time spent.
- We incur extra complexity handling multiple dataloaders inside each loop since we also need to account for dataloader_idx as another level of nesting (across control flow, logging, and metrics)

We can have abstract classes for unifying: 
**`DataIterationStrategy`**: 
- indicates how the dataloaders are iterated over.
- *Context*: 
   - Users want to customize their data iteration strategy in addition to current strategy that PL has provided (see below, it’s iterating the dataloaders in order). For example, multi-task learning may require loading data from a single task or multiple tasks in a certain way. With pre-defined strategy, users can define the training/validation datasets efficiently. 
- *Attributes*:
   - This varies by strategy
- *Examples*:
   - InOrder 
      - this is current default iterator strategy in Pytorch Lightning for multiple dataloaders in val/test loop

**`MultiIterator`**: 
```python
class MultiIterator(Iterator[Dict[str, Any]]):
   def __init__(
       self,
       individual_dataloaders: Union[
           Mapping[str, DataLoader], Mapping[str, DataLoaderInterface]
       ],
       iteration_strategy: DataIterationStrategy,
   ) -> None:
       self.individual_dataloaders = individual_dataloaders
       self.iteration_strategy = iteration_strategy

   def __str__(self) -> str:
       return str(self.iteration_strategy)

   @abstractmethod
   def __next__(self) -> Dict[str, Any]:
       pass
```
- defines the iteration logic to get a batch given batches from all individual dataloaders
- *Attributes*:
   - individual dataloaders: map
   - iteration_startegy
- *Example*:
   - InOrderIterator
      - Return all batches from a single dataset until it is exhausted and then moves to the next one. 
   - RandomizedBatchSamplerIterator
      - Return randomly samples from each dataset using each dataset using the provided weights
   - RoundRobinIterator
      - Cycles over the dataloader one by one based on RoundRobin Stargey
   - AllDatasetBatchesIterator
      - Returns a dict containing batches from all dataloaders
      
**`MultiDataloader`**: 
```python
class MultiDataloader:
   def __init__(
       self,
       individual_dataloaders: Union[
           Dict[str, DataLoader], Dict[str, DataLoaderInterface]
       ],
       iteration_strategy: DataIterationStrategy,
       iterator_class: Optional[Type[MultiIterator]] = None,
       ignore_empty_data: bool = False,
   ) -> None:
       self.individual_dataloaders = individual_dataloaders
       self.iteration_strategy = iteration_strategy
       self.iterator_class = iterator_class
       self._check_empty_data(self.ignore_empty_data)

   def __iter__(self) -> Iterator[Dict[str, Any]]:
       iterator_cls = self.iterator_cls
       if iterator_cls is None:
           iterator_cls = DataIterationStrategyRegistry.get(self.iteration_strategy)
       self.iterator = iterator_cls(
           individual_dataloaders=self.individual_dataloaders,
           iteration_strategy=self.iteration_strategy,
       )
       return self.iterator

   def shutdown(self) -> None:
       pass
```
- it cycles through the individual dataloaders.
- *Attributes*:
   - Iteration_strategy
   - Iterator_class
   - ignore_empty_data

#### How to use multi dataloder?
```python
def train_dataloader(self):
    dataloader_1 = DataLoader(RandomDataset(32, 8), batch_size=8)
    dataloader_2 = DataLoader(RandomDataset(32, 16), batch_size=8)

    individual_dataloaders = {"1": dataloader_1, "2": dataloader_2}

    all_dataset_batches = AllDatasetBatches()

    multi_dataloader = iter(
           MultiDataloader(
               individual_dataloaders,
               all_dataset_batches,
           )
       )
    return next(multi_dataloader)

def training_step(self, batch, batch_idx):
    batch_1 = batch["1"]
    batch_2 = batch["2"]

```
These utilities are used by Facebook internally for handling multiple dataloaders, and they are not specific to FB infra. PyTorch Lightning users can also be benefited from this new feature.

Comparison between existing mult-dataset support from PL and the new feature

|Current      | New Feature |
| ----------- | ----------- |
| Create a DataLoader that iterates over multiple Datasets under the hood.      | N/A       |
| In the training loop you can pass multiple DataLoaders as a dict or list/tuple and Lightning will automatically combine the batches from different DataLoaders.   | MultiDataLoader with AllDatasetBatchesStrategy        |
| In the validation and test loop you have the option to return multiple DataLoaders, which Lightning will call sequentially. | MultiDataLoader with InOrderStrategy |
|CombinedLoader | MultiDataLoader with RandomizedBatchSamplerStrategy |

From this comparison, 
- we can see that this new feature could unify the way to handle multi-datasets and multi dataloders. 
- we can de-couple data handling from different training loop stage. 
- We provide a unified and flexible way for users to define or contribute their iteration strategy for data processing. 

#### Discussion about on-going work of fault tolerance training [google doc](https://docs.google.com/document/d/1pvhyNg17pm7-ECMTpgpZiJf6sbypeuaB8QsSP79n6k0/edit?usp=sharing) in PyTorch Lightning. 
Thanks for the valuable comments from @tchaton! The new experimental feature - fault tolerance training, supports collection of dataloaders with RandomizedBatchSamplerStrategy. As we are introducing a more unified approach to support multi-dataloders, we will consider how to cooperate with the current fault tolerance training feature. To keep this issue clean, I will create another issue to discuss more about this. 

### Pitch

- This will also involve expanding the data loader interface to accept an Iterable instead of collections of PyTorch dataloaders.
- Let's remove CombinedLoader from the Trainer internals. This should be supplied via the data interfaces
- Multiple dataloader mode should be a property of the data agent, not the Trainer: https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html

### Alternatives

Keep as it is. 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by ananthsub** [comment]
@ninginthecloud It'd be very helpful to include more design detail on the iteration strategy, iterator, and dataloader for people to comment on. If you think this would be readable in the issue, feel free to post there. If it would require more space, we could write this up in a separate doc and link it from this issue. What do you think?

**Comment by tchaton** [comment]
Dear @ninginthecloud,

One note to be added early there. Lightning currently supports fault tolerant training for collection of dataloaders and therefore those features should be supported for the new abstractions.

Best,
T.C

**Comment by ninginthecloud** [comment]
Hi, @ananthsub  I will include more details in a doc and add the link here.

**Comment by ninginthecloud** [comment]
Hi, @tchaton Thanks for the valuable information~ Do you have any docs/links for the current support for collection of dataloaders?  In case I'm missing something important here. I will compare the proposed features with existing features as a follow up.

**Comment by tchaton** [comment]
Dear @ninginthecloud,

We didn't create a document yet as it is still on-going, 
but here is one of the latest PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/8950.

The logic can be found under those files primarily for DataLoader:

Supporters: https://github.com/PyTorchLightning/pytorch-lightning/blob/e727d8b46d27644724ceec93d5fc7b4651a71520/pytorch_lightning/trainer/supporters.py

The entire auto_restart file: https://github.com/PyTorchLightning/pytorch-lightning/blob/e727d8b46d27644724ceec93d5fc7b4651a71520/pytorch_lightning/utilities/auto_restart.py

Tests:
test_auto_restart.py, /test_auto_restart.py

Search for `_fault_tolerant_training()` within PyTorch Lightning to find all the places where fault tolerant is added.

Best,
T.C
